degica/barcelona-cli
# Login vault

## Context

This PR introduces a mechanism to login to vault. This PR has a large number of line changes, which are mostly tests (login was not tested before) and mocks to facilitate testing. If you find this PR too long, I can split it up a little more into possibly a refactor PR and test PR.

## How to test
> Since this is a client, you don't need to have a bcn account, simply knowing that it sends the right requests is sufficient, though you are welcome to `kaiser up` barcelona on your local machine and following [these instructions](https://github.com/degica/barcelona/pull/605) if you wish to see everything working.
1. Simply attempt to login to vault by going: `bcn -d login https://example.com --auth vault`
2. You should be prompted to enter a github token and vault server url, just enter some random value. It doesn't matter. You can use "https://example.com" to test past any errors you may encounter if you use an invalid URL.
3. You should see the following result
```
Logging in with Vault
Create new GitHub access token with read:org permission here https://github.com/settings/tokens/new
GitHub Token: URL of vault server (e.g. https://vault.degica.com)
Vault server URL: https://example.com
POST /v1/auth/github/login?debug=true HTTP/1.1
Host: example.com
User-Agent: Go-http-client/1.1
Content-Length: 20
Accept: application/json
Content-Type: application/json
Accept-Encoding: gzip

{"token":"adasdasd"}
HTTP/2.0 404 Not Found
... and other errors
```
4. But observe that it says `logging in with Vault` and sends the token you provided to the vault server.
